Unhandled NumberFormatException in StringUtil#isUnsignedInteger(String)

Running the command delete `12312432535323212321` results in
>Caused by: java.lang.NumberFormatException: For input string: "12312432535323212321"

This Exception is not handled by the method StringUtil#isUnsignedInteger(String).

Identified in #423

--- a/src/main/java/seedu/address/commons/util/StringUtil.java
+++ b/src/main/java/seedu/address/commons/util/StringUtil.java
@@ -52,12 +52,20 @@ public class StringUtil {
     }
 
     /**
-     * Returns true if s represents an unsigned integer e.g. 1, 2, 3, ... <br>
-     * Will return false if the string is:
-     * null, empty string, "-1", "0", "+1", and " 2 " (untrimmed) "3 0" (contains whitespace).
-     * @param s Should be trimmed.
+     * Returns true if {@code s} represents a non-zero unsigned integer
+     * e.g. 1, 2, 3, ..., {@code Integer.MAX_VALUE} <br>
+     * Will return false for any other non-null string input
+     * e.g. empty string, "-1", "0", "+1", and " 2 " (untrimmed), "3 0" (contains whitespace), "1 a" (contains letters)
+     * @throws NullPointerException if {@code s} is null.
      */
-    public static boolean isUnsignedInteger(String s) {
-        return s != null && s.matches("^0*[1-9]\\d*$");
+    public static boolean isNonZeroUnsignedInteger(String s) {
+        requireNonNull(s);
+
+        try {
+            int value = Integer.parseInt(s);
+            return value > 0 && !s.startsWith("+"); // "+1" is successfully parsed by Integer#parseInt(String)
+        } catch (NumberFormatException nfe) {
+            return false;
+        }
     }
 }

--- a/src/main/java/seedu/address/logic/parser/ParserUtil.java
+++ b/src/main/java/seedu/address/logic/parser/ParserUtil.java
@@ -23,14 +23,16 @@ import seedu.address.model.tag.Tag;
  */
 public class ParserUtil {
 
+    public static final String MESSAGE_INVALID_INDEX = "Index is not a non-zero unsigned integer.";
+
     /**
      * Parses {@code index} into an integer and returns it. Leading and trailing whitespaces will be trimmed.
      * @throws IllegalValueException if the specified index is invalid (not non-zero unsigned integer).
      */
     public static int parseIndex(String index) throws IllegalValueException {
         String trimmedIndex = index.trim();
-        if (!StringUtil.isUnsignedInteger(trimmedIndex)) {
-            throw new IllegalValueException("Index is not a non-zero unsigned integer.");
+        if (!StringUtil.isNonZeroUnsignedInteger(trimmedIndex)) {
+            throw new IllegalValueException(MESSAGE_INVALID_INDEX);
         }
         return Integer.parseInt(trimmedIndex);
 

--- a/src/test/java/seedu/address/commons/util/StringUtilTest.java
+++ b/src/test/java/seedu/address/commons/util/StringUtilTest.java
@@ -22,31 +22,34 @@ public class StringUtilTest {
     @Test
     public void isUnsignedPositiveInteger() {
 
-        // Equivalence partition: null
-        assertFalse(StringUtil.isUnsignedInteger(null));
-
         // EP: empty strings
-        assertFalse(StringUtil.isUnsignedInteger("")); // Boundary value
-        assertFalse(StringUtil.isUnsignedInteger("  "));
+        assertFalse(StringUtil.isNonZeroUnsignedInteger("")); // Boundary value
+        assertFalse(StringUtil.isNonZeroUnsignedInteger("  "));
 
         // EP: not a number
-        assertFalse(StringUtil.isUnsignedInteger("a"));
-        assertFalse(StringUtil.isUnsignedInteger("aaa"));
+        assertFalse(StringUtil.isNonZeroUnsignedInteger("a"));
+        assertFalse(StringUtil.isNonZeroUnsignedInteger("aaa"));
 
         // EP: zero
-        assertFalse(StringUtil.isUnsignedInteger("0"));
+        assertFalse(StringUtil.isNonZeroUnsignedInteger("0"));
+
+        // EP: zero as prefix
+        assertTrue(StringUtil.isNonZeroUnsignedInteger("01"));
 
         // EP: signed numbers
-        assertFalse(StringUtil.isUnsignedInteger("-1"));
-        assertFalse(StringUtil.isUnsignedInteger("+1"));
+        assertFalse(StringUtil.isNonZeroUnsignedInteger("-1"));
+        assertFalse(StringUtil.isNonZeroUnsignedInteger("+1"));
 
         // EP: numbers with white space
-        assertFalse(StringUtil.isUnsignedInteger(" 10 ")); // Leading/trailing spaces
-        assertFalse(StringUtil.isUnsignedInteger("1 0"));  // Spaces in the middle
+        assertFalse(StringUtil.isNonZeroUnsignedInteger(" 10 ")); // Leading/trailing spaces
+        assertFalse(StringUtil.isNonZeroUnsignedInteger("1 0"));  // Spaces in the middle
+
+        // EP: number larger than Integer.MAX_VALUE
+        assertFalse(StringUtil.isNonZeroUnsignedInteger(Long.toString(Integer.MAX_VALUE + 1)));
 
         // EP: valid numbers, should return true
-        assertTrue(StringUtil.isUnsignedInteger("1")); // Boundary value
-        assertTrue(StringUtil.isUnsignedInteger("10"));
+        assertTrue(StringUtil.isNonZeroUnsignedInteger("1")); // Boundary value
+        assertTrue(StringUtil.isNonZeroUnsignedInteger("10"));
     }
 
 

--- a/src/test/java/seedu/address/logic/parser/ParserUtilTest.java
+++ b/src/test/java/seedu/address/logic/parser/ParserUtilTest.java
@@ -3,6 +3,7 @@ package seedu.address.logic.parser;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
 import static org.junit.Assert.assertTrue;
+import static seedu.address.logic.parser.ParserUtil.MESSAGE_INVALID_INDEX;
 
 import java.util.Arrays;
 import java.util.Collections;
@@ -45,9 +46,10 @@ public class ParserUtilTest {
     }
 
     @Test
-    public void parseIndex_outOfRangeInput_throwsNumberFormatException() throws Exception {
-        thrown.expect(NumberFormatException.class);
-        ParserUtil.parseIndex("3423423423423432");
+    public void parseIndex_outOfRangeInput_throwsIllegalValueException() throws Exception {
+        thrown.expect(IllegalValueException.class);
+        thrown.expectMessage(MESSAGE_INVALID_INDEX);
+        ParserUtil.parseIndex(Long.toString(Integer.MAX_VALUE + 1));
     }
 
     @Test

